soc: telink: rcp: add intermediate transport layer

Nazarii-Bodnar opened the pull request:
Add wrapper layer for the RCP communication
protocol to provide multiple options instead of
hardcoded UART.

Nazarii-Bodnar committed:
soc: telink: rcp: add intermediate transport layer

Add wrapper layer for the RCP communication
protocol to provide multiple options instead of
hardcoded UART.

Signed-off-by: Nazarii Bodnar <nazarii.bodnar@telink-semi.com>

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/Kconfig.ot_rcp:
```diff
@@ -16,6 +16,10 @@ if TELINK_OT_RCP
 config OPENTHREAD_RADIO_WORKQUEUE_STACK_SIZE
 	default 2048
 
+config TELINK_OT_RCP_UART
```
Please create choice here. For now add dummy to check compilation.

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/ot_rcp.h:
```diff
@@ -23,7 +24,7 @@ struct openthread_rcp_buffer {
 
 struct openthread_rcp_data {
-	const struct device *uart;
+	const struct device *rcp_transport_dev;
```
Should be transport instance here not a pointer.

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/ot_rcp.h:
```diff
@@ -23,7 +24,7 @@ struct openthread_rcp_buffer {
+	const struct device *rcp_transport_dev;
 	struct k_work work;
 	uint8_t rb_data[CONFIG_TELINK_OT_RCP_BUFFER_SIZE];
 	struct ring_buf rb;
```
I think it should be moved to transport. Check all structure carefully.

andriy-bilynskyy replied:
There is no changes here...
struct k_work work;
uint8_t rb_data[CONFIG_TELINK_OT_RCP_BUFFER_SIZE];
struct ring_buf rb;
it's UART only relative stuffs and should be moved to transport class.

Nazarii-Bodnar committed:
soc: telink: rcp: rework transport wrapper structure

Change transport wrapper structure, modify transport instance
and rework transport interrupt handling.

Signed-off-by: Nazarii Bodnar <nazarii.bodnar@telink-semi.com>

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/Kconfig.ot_rcp:
```diff
@@ -16,6 +16,15 @@ if TELINK_OT_RCP
+
+	config TELINK_OT_RCP_UART
+		bool "Set UART as a transport for the OT RCP"
+
```
Please add dummy for demonstration.
config TELINK_OT_RCP_TRANSPORT_DUMMY
and some dummy back-end implementation c,h files

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/Kconfig.ot_rcp:
```diff
@@ -16,6 +16,15 @@ if TELINK_OT_RCP
+	bool "Transport for the OT RCP"
+	default TELINK_OT_RCP_UART
+
+	config TELINK_OT_RCP_UART
```
Please keep base choice here:
config TELINK_OT_RCP_TRANSPORT_UART

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/ot_rcp.c:
```diff
@@ -40,35 +39,6 @@ SYS_INIT(openthread_rcp_work_q_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DE
-	}
-}
-
 static void openthread_rcp_reception_work(struct k_work *item)
```
also move to transport

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/transport/rcp_transport.h:
```diff
@@ -0,0 +1,37 @@
+#include <zephyr/device.h>
+
+struct rcp_transport_data {
+#ifdef CONFIG_TELINK_OT_RCP_UART
```
No special compilation chooses here.
Dummy backed should provide same interface. So just change c,h files in CMakelist.

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/transport/rcp_transport.h:
```diff
@@ -0,0 +1,37 @@
+typedef void (*rcp_transport_irq_t)(struct rcp_transport_data rcp_transport);
+
+int rcp_transport_init(struct rcp_transport_data *rcp_transport);
+int rcp_transport_deinit(struct rcp_transport_data rcp_transport);
```
keep pointers. in no changes const pointer.
that's for all functions.

Nazarii-Bodnar committed:
soc: telink: rcp: add dummy transport

Add dummy interface to the ot rcp transport wrapper.

Signed-off-by: Nazarii Bodnar <nazarii.bodnar@telink-semi.com>

andriy-bilynskyy commented on soc/telink/tlsr/common/ot_rcp/transport/rcp_transport.h:
```diff
@@ -0,0 +1,48 @@
+void rcp_transport_reception_handler_set(struct rcp_transport_data *rcp_transport,
+					 rpc_transport_reception_t rcp_transport_reception_handler);
+
+void rcp_transport_irq_enable(const struct rcp_transport_data *rcp_transport);
```
It shouldn't be an IRQ.
Init(on byte received, on rx complete)
Deinit
OutOneByte
TransmittBuffer

andriy-bilynskyy commented:
please, squash commits before merging
